transformer_engine/pytorch/module/base.py

```diff
@@ -88,14 +90,53 @@ def initialize_ub(
     global _ub_communicators
     assert _ub_communicators is None, "UB communicators are already initialized."
     _ub_communicators = {}
-    rank_id = torch.distributed.get_rank()
-    world_size = torch.distributed.get_world_size()
-    tp_id = torch.distributed.get_rank(tp_group)
-    tp_size = torch.distributed.get_world_size(tp_group)
+    world_group = torch.distributed.new_group(backend="nccl")
+    world_rank = torch.distributed.get_rank(world_group)
+    world_size = torch.distributed.get_world_size(world_group)
+
+    # Construct an intra-node communicator -- this should include ALL ranks in the node
+    # NOTE: This may be different than the tensor-parallel group (e.g. two TP groups in a node),
+    #       in which case the local_size we get below will not be equal to the tp_size given
+    #       by the user. Userbuffers internally accounts for this.
+    hostnames = [None for _ in range(world_size)]
+    hostname = socket.gethostname()
+    torch.distributed.all_gather_object(hostnames, hostname)
+    intra_node_ranks = []
+    for i, host in enumerate(hostnames):
+        if host == hostname:
+            intra_node_ranks.append(i)
+    if len(intra_node_ranks) == world_size:
+        intra_node_group = world_group
+        local_rank = world_rank
+        local_size = world_size
+    else:
+        intra_node_group = torch.distributed.new_group(backend="nccl", ranks=intra_node_ranks)
+        local_rank = torch.distributed.get_rank(intra_node_group)
+        local_size = torch.distributed.get_world_size(intra_node_group)
+
+    # Determine node ID as the row index of the world rank in a 2D device mesh that we construct
+    # from all-gathered intra-node (local) ranks -- this avoids any assumptions about world rank
+    # ordering within the node.
+    num_nodes = world_size // local_size
+    if num_nodes > 1:
+        intra_node_ranks = torch.tensor(intra_node_ranks, dtype=torch.int64, device="cuda")
+        all_intra_node_ranks = torch.zeros(world_size, local_size, dtype=torch.int64, device="cuda")
+        torch.distributed.all_gather_into_tensor(all_intra_node_ranks, intra_node_ranks)
+        device_mesh = []
+        for rank in range(world_size):
+            node = all_intra_node_ranks[rank, :].tolist()
+            if node not in device_mesh:
+                device_mesh.append(node)
+        assert len(device_mesh) == num_nodes, "Internal TE error: Could not infer device mesh."
+        device_mesh = torch.tensor(device_mesh, dtype=int, device="cpu")
+        rank_idx = (device_mesh == world_rank).nonzero()
+        node_id = rank_idx[0, 0].item()
+    else:
+        node_id = 0
 
     # Increase the workspace by the number of maximum concurrent streams
     global _cublas_workspace
-    _cublas_workspace = get_workspace().repeat(_NUM_MAX_UB_STREAMS)
```

Comment:
(Minor point) `_NUM_MAX_UB_STREAMS` is now dangling, and could be removed from this file at line 49